File: server_py/haversine.py
"""
Haversine 공식 — 두 GPS 좌표 간 거리 계산 (미터 단위)
GPS 조작 방지를 위해 서버 측에서 재검증에 사용
"""
import math


def haversine(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    """두 좌표 사이의 거리를 미터(m) 단위로 반환"""
    R = 6_371_000  # 지구 반지름(m)

    phi1  = math.radians(lat1)
    phi2  = math.radians(lat2)
    dphi  = math.radians(lat2 - lat1)
    dlam  = math.radians(lng2 - lng1)

    a = math.sin(dphi / 2) ** 2 + math.cos(phi1) * math.cos(phi2) * math.sin(dlam / 2) ** 2
    return R * 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))

File: server/haversine.js
/**
 * Haversine 공식 — 두 GPS 좌표 간 거리 계산 (미터 단위)
 */
function haversine(lat1, lng1, lat2, lng2) {
  const R    = 6_371_000; // 지구 반지름(m)
  const toRad = deg => (deg * Math.PI) / 180;
  const dLat = toRad(lat2 - lat1);
  const dLng = toRad(lng2 - lng1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLng / 2) ** 2;
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

module.exports = { haversine };
